README
POE-Component-Client-eris

POE Session for the eris event correlation engine.

INSTALLATION

To install this module, run the following commands:

    perl Makefile.PL
    make
    make test
    make install

To install the server component, you need to hook into 
syslog-ng, add this line to your /etc/syslog-ng/syslog-ng.conf

	destination d_eris { program("PERLBIN/eris-syslog-ng.pl"); };

Replacing PERLBIN for where 'make install' stashed eris-syslog-ng.pl.

Also, send data to it:

	log { source(s_sys); source(s_tcp); source(s_udp); filter(f_whatever); destination(d_eris); };

Restart syslog-ng, and you should be able to telnet to localhost:9514


SUPPORT AND DOCUMENTATION

After installing, you can find documentation for this module with the perldoc command.

    perldoc POE::Component::Client::eris

You can also look for information at:

    Search CPAN
        http://search.cpan.org/dist/POE-Component-Client-eris

    CPAN Request Tracker:
        http://rt.cpan.org/NoAuth/Bugs.html?Dist=POE-Component-Client-eris

    AnnoCPAN, annotated CPAN documentation:
        http://annocpan.org/dist/POE-Component-Client-eris

    CPAN Ratings:
        http://cpanratings.perl.org/d/POE-Component-Client-eris

COPYRIGHT AND LICENCE

Copyright (C) 2007 Brad Lhotsky

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.


Changes
Revision history for POE-Component-Client-eris
1.2     2012-11-08 Brad Lhotsky <brad@divisionbyzero.net>
        - Added 'ReturnType' option to skip Parse::Syslog::Line

1.1     2012-10-15 Brad Lhotsky <brad@divisionbyzero.net>
        - Fixed Subscribe/Match logic

1.0     2012-09-07 Brad Lhotsky <brad@divisionbyzero.net>
        - Wrapped parse_syslog_line in eval {};

0.9     2012-08-23 Brad Lhotsky <brad@divisionbyzero.net>
        - Removed taint checking in tests
        - Updated author email

0.8     2012-07-06 Brad Lhotsky <brad@divisionbyzero.net>
        - Fixed bug in match handling

0.7     2012-06-28 Brad Lhotsky <brad@divisionbyzero.net>
        - Moved to Dist::Zilla
        - Fixed whitespace at end of lines

0.6 	2011-07-08 Brad Lhotsky <brad@divisionbyzero.net>
		- Added to https://github.com/reyjrar/POE-Component-Client-eris
		- Added to PAUSE
		- scripts/eris-syslog-ng.pl included in distribution
		- supports match and subscription

0.01    2007-03-02 Brad Lhotsky <brad@divisionbyzero.net>
        First version, only basic functionality



scripts/eris-syslog-ng.pl
#!/usr/bin/env perl
#
# This is the POE Master Server.
#  1) Take all the syslog input
#  2) Listen for parsers
#  3) Filter streams to parsers

use strict;
use warnings;

use Socket;

# POE System
use POE qw(
	Wheel::ReadWrite
	Component::Server::TCP
);


my %cooked = (
	program => qr/\s+\d+:\d+:\d+\s+\S+\s+([^:\s]+)(:|\s)/,
);

#--------------------------------------------------------------------------#
# POE Session Initialization


# Dispatcher Master Session
POE::Session->create(
	inline_states => {
		_start					=> \&dispatcher_start,
		_stop					=> sub { print "SESSION ", $_[SESSION]->ID, " stopped.\n"; },
		register_client			=> \&register_client,
		subscribe_client		=> \&subscribe_client,
		unsubscribe_client		=> \&unsubscribe_client,
		fullfeed_client			=> \&fullfeed_client,
		dispatch_message		=> \&dispatch_message,
		broadcast				=> \&broadcast,
		hangup_client			=> \&hangup_client,
		server_shutdown			=> \&server_shutdown,
		match_client			=> \&match_client,
		nomatch_client			=> \&nomatch_client,
		debug_client			=> \&debug_client,
		nobug_client			=> \&nobug_client,
		debug_message			=> \&debug_message,
	},
);

# TCP Session Master
POE::Component::Server::TCP->new(
		Alias		=> 'server',
		Address		=> '127.0.0.1',
		Port		=> 9514,

		ClientConnected		=> \&client_connect,
		ClientInput			=> \&client_input,

		ClientDisconnected	=> \&client_term,
		ClientError			=> \&client_term,

		InlineStates		=> {
			client_print		=> \&client_print,
		},
);

# Syslog-ng Stream Master
POE::Session->create(
		inline_states => {
			_start		=> \&stream_start,
			_stop		=> sub { print "SESSION ", $_[SESSION]->ID, " stopped.\n"; },

			stream_line		=> \&stream_line,
			stream_error	=> \&stream_error,
		},
);

#--------------------------------------------------------------------------#

#--------------------------------------------------------------------------#
# POE Main Loop
POE::Kernel->run();
exit 0;
#--------------------------------------------------------------------------#


#--------------------------------------------------------------------------#
# POE Event Functions
#--------------------------------------------------------------------------#

#--------------------------------------------------------------------------#
sub debug {
	my $msg = shift;
	chomp($msg);
	$poe_kernel->post( 'dispatcher' => 'debug_message' => $msg );
	print "[debug] $msg\n";
}
#--------------------------------------------------------------------------#
sub dispatcher_start {
	my ($kernel, $heap) = @_[KERNEL, HEAP];

	$kernel->alias_set( 'dispatcher' );

	$heap->{subscribers} = { };
	$heap->{full} = { };
	$heap->{debug} = { };
	$heap->{match} = { };
}

#--------------------------------------------------------------------------#
sub register_client {
	my ($kernel,$heap,$sid) = @_[KERNEL,HEAP,ARG0];

	$heap->{clients}{$sid} = 1;
}

#--------------------------------------------------------------------------#
sub debug_client {
	my ($kernel,$heap,$sid) = @_[KERNEL,HEAP,ARG0];

	if( exists $heap->{full}{$sid} ) {  return;  }

	$heap->{debug}{$sid} = 1;
	$kernel->post( $sid => 'client_print' => 'Debugging enabled.' );
}

#--------------------------------------------------------------------------#
sub nobug_client {
	my ($kernel,$heap,$sid) = @_[KERNEL,HEAP,ARG0];

	delete $heap->{debug}{$sid}
		if exists $heap->{debug}{$sid};
	$kernel->post( $sid => 'client_print' => 'Debugging disabled.' );
}

#--------------------------------------------------------------------------#
sub fullfeed_client {
	my ($kernel,$heap,$sid) = @_[KERNEL,HEAP,ARG0];

	#
	# Remove from normal subscribers.
	foreach my $prog (keys %{ $heap->{subscribers} }) {
		delete $heap->{subscribers}{$prog}{$sid}
			if exists $heap->{subscribers}{$prog}{$sid};
	}

	#
	# Turn off DEBUG
	if( exists $heap->{debug}{$sid} ) {
		delete $heap->{debug}{$sid};
	}

	#
	# Add to fullfeed:
	$heap->{full}{$sid} = 1;

	$kernel->post( $sid => 'client_print' => 'Full feed enabled, all other functions disabled.');
}

#--------------------------------------------------------------------------#
sub subscribe_client {
	my ($kernel,$heap,$sid,$argstr) = @_[KERNEL,HEAP,ARG0,ARG1];

	if( exists $heap->{full}{$sid} ) {  return;  }

	my @progs = map { lc } split /[\s,]+/, $argstr;
	foreach my $prog (@progs) {
		$heap->{subscribers}{$prog}{$sid} = 1;
	}

	$kernel->post( $sid => 'client_print' => 'Subscribed to : ' . join(', ', @progs ) );
}
#--------------------------------------------------------------------------#
sub unsubscribe_client {
	my ($kernel,$heap,$sid,$argstr) = @_[KERNEL,HEAP,ARG0,ARG1];

	my @progs = map { lc } split /[\s,]+/, $argstr;
	foreach my $prog (@progs) {
		delete $heap->{subscribers}{$prog}{$sid};
	}

	$kernel->post( $sid => 'client_print' => 'Subscription removed for : ' . join(', ', @progs ) );
}

#--------------------------------------------------------------------------#
sub match_client {
	my ($kernel,$heap,$sid,$argstr) = @_[KERNEL,HEAP,ARG0,ARG1];

	if( exists $heap->{full}{$sid} ) {  return;  }

	my @words = map { lc } split /[\s,]+/, $argstr;
	foreach my $word (@words) {
		$heap->{match}{$word}{$sid} = 1;
	}

	$kernel->post( $sid => 'client_print' => 'Receiving messages matching : ' . join(', ', @words ) );
}
#--------------------------------------------------------------------------#
sub nomatch_client {
	my ($kernel,$heap,$sid,$argstr) = @_[KERNEL,HEAP,ARG0,ARG1];

	my @words = map { lc } split /[\s,]+/, $argstr;
	foreach my $word (@words) {
		delete $heap->{match}{$word}{$sid};
		# Remove the word from searching if this was the last client
		delete $heap->{match}{$word} unless keys %{ $heap->{match}{$word} };
	}


	$kernel->post( $sid => 'client_print' => 'No longer receving messages matching : ' . join(', ', @words ) );
}
#--------------------------------------------------------------------------#
sub hangup_client {
	my ($kernel,$heap,$sid) = @_[KERNEL,HEAP,ARG0];

	delete $heap->{clients}{$sid};

	foreach my $p ( keys %{ $heap->{subscribers} } ) {
		delete $heap->{subscribers}{$p}{$sid}
			if exists $heap->{subscribers}{$p}{$sid};
	}

	foreach my $word ( keys %{ $heap->{match} } ) {
		delete $heap->{match}{$word}{$sid}
			if exists $heap->{match}{$word}{$sid};
		# Remove the word from searching if this was the last client
		delete $heap->{match}{$word} unless keys %{ $heap->{match}{$word} };
	}


	if( exists $heap->{debug}{$sid} ) {
		delete $heap->{debug}{$sid};
	}

	if( exists $heap->{full}{$sid} ) {
		delete $heap->{full}{$sid};
	}

	debug("Client Termination Posted: $sid\n");

}


#--------------------------------------------------------------------------#
sub stream_start {
	my ($kernel, $heap) = @_[KERNEL, HEAP];

	$kernel->alias_set( 'stream' );

	#
	# Initialize the connection to STDIN as a POE::Wheel
	my $stdin = IO::Handle->new_from_fd( \*STDIN, 'r' );
	my $stderr = IO::Handle->new_from_fd( \*STDERR, 'w' );

	$heap->{stream} = POE::Wheel::ReadWrite->new(
		InputHandle		=> $stdin,
		OutputHandle	=> $stderr,
		InputEvent		=> 'stream_line',
		ErrorEvent		=> 'stream_error',
	);
}


#--------------------------------------------------------------------------#
sub stream_line {
	my ($kernel,$msg) = @_[KERNEL,ARG0];

	return unless length $msg;

	$kernel->post( 'dispatcher' => 'dispatch_message' => $msg );

}

#--------------------------------------------------------------------------#
sub stream_error {
	my ($kernel) = $_[KERNEL];

	debug("STREAM ERROR!!!!!!!!!!\n");
	$kernel->call( 'dispatcher' => 'server_shutdown' => 'Stream lost' );
}



#--------------------------------------------------------------------------#
sub server_shutdown {
	my ($kernel,$heap,$msg) = @_[KERNEL,HEAP,ARG0];

	$kernel->call( dispatcher => 'broadcast' => 'SERVER DISCONNECTING: ' . $msg );
	$kernel->call( 'server' => 'shutdown' );
	exit;
}


#--------------------------------------------------------------------------#
sub client_connect {
	my ($kernel,$heap,$ses) = @_[KERNEL,HEAP,SESSION];

	my $KID = $kernel->ID();
	my $CID = $heap->{client}->ID;
	my $SID = $ses->ID;

	$kernel->post( 'dispatcher' => 'register_client' => $SID );

	$heap->{clients}{ $SID } = $heap->{client};
	#
	# Say hello to the client.
	$heap->{client}->put( "EHLO Streamer (KERNEL: $KID:$SID)" );
}

#--------------------------------------------------------------------------#
sub client_print {
	my ($kernel,$heap,$ses,$mesg) = @_[KERNEL,HEAP,SESSION,ARG0];

	$heap->{clients}{$ses->ID}->put($mesg);
}

#--------------------------------------------------------------------------#
sub broadcast {
	my ($kernel,$heap,$msg) = @_[KERNEL,HEAP,ARG0];

	foreach my $sid (keys %{ $heap->{clients} }) {
		$kernel->post( $sid => 'client_print' => $msg );
	}
}
#--------------------------------------------------------------------------#
sub dispatch_message {
	my ($kernel,$heap,$msg) = @_[KERNEL,HEAP,ARG0];

	foreach my $sid ( keys %{ $heap->{full} } ) {
		$kernel->post( $sid => 'client_print' => $msg );
	}

	# Program based subscriptions
	if( my ($program) = map { lc } ($msg =~ /$cooked{program}/) ) {
		# remove the sub process and PID from the program
		$program =~ s/\(.*//g;
		$program =~ s/\[.*//g;

		debug("DISPATCHING MESSAGE [$program]");

		if( exists $heap->{subscribers}{$program} ) {
			foreach my $sid (keys %{ $heap->{subscribers}{$program} }) {
				$kernel->post( $sid => 'client_print' => $msg );
			}
		}
		else {
			debug("Message discarded, no listeners.");
		}
	}

	# Match based subscriptions
	if( keys %{ $heap->{match} } ) {
		foreach my $word (keys %{ $heap->{match} } ) {
			if( index( $msg, $word ) != -1 ) {
				foreach my $sid ( keys %{ $heap->{match}{$word} } ) {
					$kernel->post( $sid => 'client_print' => $msg );
				}
			}
		}
	}
}

#--------------------------------------------------------------------------#
sub debug_message {
	my ($kernel,$heap,$msg) = @_[KERNEL,HEAP,ARG0];


	foreach my $sid (keys %{ $heap->{debug} }) {
		$kernel->post( $sid => 'client_print' => '[debug] ' . $msg );
	}
}


#--------------------------------------------------------------------------#
sub client_input {
	my ($kernel,$heap,$ses,$msg) = @_[KERNEL,HEAP,SESSION,ARG0];
	my $sid = $ses->ID;

	if( !exists $heap->{dispatch}{$sid} ) {
		$heap->{dispatch}{$sid} = {
			fullfeed		=> {
				re			=> qr/^(fullfeed)/,
				callback	=> sub {
					$kernel->post( 'dispatcher' => 'fullfeed_client' => $sid );
				},
			},
			subscribe		=> {
				re			=> qr/^sub(?:scribe)? (.*)/,
				callback	=> sub {
					$kernel->post( 'dispatcher' => 'subscribe_client' => $sid, shift );
				},
			},
			unsubscribe 	=> {
				re			=> qr/^unsub(?:scribe)? (.*)/,
				callback	=> sub {
					$kernel->post( 'dispatcher' => 'unsubscribe_client' => $sid, shift );
				},
			},
			match 	=> {
				re			=> qr/^match (.*)/i,
				callback	=> sub {
					$kernel->post( 'dispatcher' => 'match_client' => $sid, shift );
				},
			},
			nomatch 	=> {
				re			=> qr/^nomatch (.*)/i,
				callback	=> sub {
					$kernel->post( 'dispatcher' => 'nomatch_client' => $sid, shift );
				},
			},
			debug 	=> {
				re			=> qr/^(debug)/i,
				callback	=> sub {
					$kernel->post( 'dispatcher' => 'debug_client' => $sid, shift );
				},
			},
			nobug 	=> {
				re			=> qr/^(no(de)?bug)/i,
				callback	=> sub {
					$kernel->post( 'dispatcher' => 'nobug_client' => $sid, shift );
				},
			},
			#quit			=> {
			#	re			=> qr/(exit)|q(uit)?/,
			#	callback	=> sub {
			#			$kernel->post( $sid => 'client_print' => 'Terminating connection on your request.');
			#			$kernel->post( $sid => 'shutdown' );
			#	},
			#},
			#status			=> {
			#	re			=> qr/^status/,
			#	callback	=> sub {
			#		my $cnt = scalar( keys %{ $heap->{clients} } );
			#		my $subcnt = scalar( keys %{ $heap->{subscribers} });
			#		my $msg = "Currently $cnt connections, $subcnt subscribed.";
			#		$kernel->post( $sid, 'client_print', $msg );
			#	},
			#},
		};
	}

	#
	# Check for messages:
	my $handled = 0;
	my $dispatch = $heap->{dispatch}{$sid};
	foreach my $evt ( keys %{ $dispatch } ) {
		if( my($args) = ($msg =~ /$dispatch->{$evt}{re}/)) {
			$handled = 1;
			$dispatch->{$evt}{callback}->($args);
			last;
		}
	}

	if( !$handled ) {
		$kernel->post( $sid => 'client_print' => 'UNKNOWN COMMAND, Ignored.' );
	}
}

#--------------------------------------------------------------------------#
sub client_term {
	my ($kernel,$heap,$ses) = @_[KERNEL,HEAP,SESSION];
	my $sid = $ses->ID;

	delete $heap->{dispatch}{$sid};
	$kernel->post( 'dispatcher' => 'hangup_client' =>  $sid );

	debug("SERVER, client $sid disconnected.\n");
}


#--------------------------------------------------------------------------#


lib/POE/Component/Client/eris.pm
# ABSTRACT: POE Component for reading eris events

package POE::Component::Client::eris;

use warnings;
use strict;
use Carp;
use Parse::Syslog::Line;

use POE qw(
    Component::Client::TCP
);

our $VERSION = '1.2';

=head1 SYNOPSIS

POE session for integration with the eris event correlation engine.

    use POE::Component::Client::eris;

    my $eris_sess_id = POE::Component::Client::eris->spawn(
            RemoteAddress       => 'localhost',     #default
            RemotePort          => '9514',          #default
            Alias               => 'eris_client',   #default
            Subscribe           => [qw(snort dhcpd)],               # REQUIRED (and/or Match)
            Match               => [qw(devbox1 myusername error)],  # REQUIRED (and/or Subscribe)
            MessageHandler      => sub { ... },      # REQUIRED
    );
    ...
    POE::Kernel->run();

For use with a server running the POE::Component::Server::eris output.

=head1 EXPORT

POE::Component::Client::eris does not export any symbols.

=head1 FUNCTIONS

=head2 spawn

Creates the POE::Session for the eris correlator.

Parameters:
    RemoteAddress       => 'localhost',     #default
    RemotePort          => '9514',          #default
    Alias               => 'eris_client',   #default
    Subscribe           => [qw(snort dhcpd)],               # REQUIRED (and/or Match)
    Match               => [qw(devbox1 myusername error)],  # REQUIRED (and/or Subscribe)
    MessageHandler      => sub { ... },      # REQUIRED
    ReturnType          => 'hash',           # default, or 'string'

=cut

sub spawn {
    my $type = shift;

    #
    # Param Setup
    my %args = (
        RemoteAddress   => 'localhost',
        RemotePort      => 9514,
        Alias           => 'eris_client',
        ReturnType      => 'hash',
        Subscribe       => undef,
        Match           => undef,
        MessageHandler  => undef,
        @_
    );

    #
    # Build the client connection
    my $tcp_sessid = POE::Component::Client::TCP->new(
        Alias           => $args{Alias},
        RemoteAddress   => $args{RemoteAddress},
        RemotePort      => $args{RemotePort},
        Filter          => 'POE::Filter::Line',
        Connected       => sub {
            my ($kernel,$heap) = @_[KERNEL,HEAP];
            $heap->{readyState} = 0;
            $heap->{connected} = 0;
            $kernel->delay( 'do_setup_pipe' => 1 );
        },
        ConnectError    => sub {
            my ($kernel,$syscall,$errid,$errstr) = @_[KERNEL,ARG0,ARG1,ARG2];
            carp "Connection Error ($errid) at $syscall: $errstr\n";
            $kernel->delay('reconnect' => 10);
        },
        Disconnected    => sub {
            my ($kernel,$heap) = @_[KERNEL,HEAP];
            $kernel->delay('reconnect' => 10);
        },
        ServerError     => sub  {
            my ($kernel,$syscall,$errid,$errstr) = @_[KERNEL,ARG0,ARG1,ARG2];
            carp "Server Error ($errid) at $syscall: $errstr\n";
            $kernel->delay('reconnect' => 5);
        },
        #
        # Handle messages from the server.
        #  Set readyState = 1 if applicable
        #  Call the inline states:
        #   handle_message (successful)
        #   handle_unknown (out of order input)
        ServerInput     => sub {
            my ($kernel,$heap,$instr) = @_[KERNEL,HEAP,ARG0];
            chomp $instr;
            if( $heap->{readyState} == 1 ) {
                $kernel->yield('handle_message' => $instr);
            }
            elsif( $heap->{connected} == 1 ) {
                if( $instr =~ /^Subscribed to \:/ ) {
                    $heap->{readyState} = 1;
                }
                elsif( $instr =~ /^Receiving / )  {
                    $heap->{readyState} = 1;
                }
                elsif( $instr =~ /^Full feed enabled/ )  {
                    $heap->{readyState} = 1;
                }
                else {
                    $kernel->yield( 'handle_unknown' => $instr );
                }
            }
            elsif( $instr =~ /^EHLO Streamer/ ) {
                $heap->{connected} = 1;
            }
            else {
                $kernel->yield( 'handle_unknown' => $instr );
            }
        },
        #
        # Inline States
        InlineStates => {
            do_setup_pipe   => sub {
                my ($kernel,$heap) = @_[KERNEL,HEAP];

                # Parse for Subscriptions or Matches
                my %data = ();
                foreach my $target (qw(Subscribe Match)) {
                    if( exists $args{$target} && defined $args{$target} ) {
                        my @data = ref $args{$target} eq 'ARRAY' ? @{ $args{$target} } : $args{$target};
                        @data = map { lc } @data if $target eq 'Subscribe';
                        next unless scalar @data > 0;
                        $data{$target} = \@data;
                    }
                }

                # Check to make sure we're doing something
                croak "Must specify a subscription or a match parameter!\n" unless keys %data;

                # Send the Subscription
                foreach my $target (sort { $a cmp $b } keys %data) {
                    my $subname = "do_" . lc $target;
                    $kernel->yield( $subname => $data{$target} );
                }
            },
            do_subscribe    => sub {
                my ($kernel,$heap,$subs) = @_[KERNEL,HEAP,ARG0];

                if( grep /^fullfeed$/, @{ $subs } ) {
                    $heap->{server}->put('fullfeed');
                }
                else {
                    $heap->{server}->put('sub ' . join(', ', @{ $subs }) );
                }
            },
            do_match    => sub {
                my ($kernel,$heap,$matches) = @_[KERNEL,HEAP,ARG0];

                $heap->{server}->put('match ' . join(', ', @{ $matches }) );
            },
            handle_message  => sub {
                my ($kernel,$heap,$instr) = @_[KERNEL,HEAP,ARG0];

                my $msg = undef;
                if( $args{ReturnType} eq 'string' ) {
                    $msg = $instr;
                }
                else {
                    eval {
                        no warnings;
                        $msg = parse_syslog_line($instr);
                    };
                    if($@ || !defined $msg) {
                        return;
                    }
                }

                if( ref $args{MessageHandler} ne 'CODE' ) {
                    croak "You need to specify a subroutine reference to the 'MessageHandler' parameter.\n";
                }
                # Try the Message Handler, eventually we can do statistics here.
                eval {
                    $args{MessageHandler}->( $msg );
                };
            },
            handle_unknown  => sub {
                my ($kernel,$heap,$msg) = @_[KERNEL,HEAP,ARG0];

                carp "UNKNOWN INPUT: $msg\n";
            },
        },
    );

    #
    # Return the TCP Session ID
    return $tcp_sessid;
}

=head1 AUTHOR

Brad Lhotsky, C<< <brad.lhotsky at gmail.com> >>

=head1 BUGS

Please report any bugs or feature requests to
C<bug-poe-component-client-eris at rt.cpan.org>, or through the web interface at
L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=POE-Component-Client-eris>.
I will be notified, and then you'll automatically be notified of progress on
your bug as I make changes.

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc POE::Component::Client::eris

You can also look for information at:

=over 4

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/POE-Component-Client-eris>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/POE-Component-Client-eris>

=item * RT: CPAN's request tracker

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=POE-Component-Client-eris>

=item * Search CPAN

L<http://search.cpan.org/dist/POE-Component-Client-eris>

=item * See also

L<http://search.cpan.org/dist/POE-Component-Server-eris>

=back

=head1 ACKNOWLEDGEMENTS

=head1 COPYRIGHT & LICENSE

Copyright 2007 Brad Lhotsky, all rights reserved.

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.

=cut

1; # End of POE::Component::Client::eris
